File: binding/src/libqregularexpression.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qregularexpression.html
pub const qregularexpression = struct {

    /// New constructs a new QRegularExpression object.
    pub fn New() QtC.QRegularExpression {
        return qtc.QRegularExpression_new();
    }


    /// New2 constructs a new QRegularExpression object.
    pub fn New2(pattern: []const u8) QtC.QRegularExpression {
        const pattern_str = qtc.libqt_string{
    .len = pattern.len,
    .data = pattern.ptr,
};
return qtc.QRegularExpression_new2(pattern_str);
    }


    /// New3 constructs a new QRegularExpression object.
    pub fn New3(re: ?*anyopaque) QtC.QRegularExpression {
        return qtc.QRegularExpression_new3(@ptrCast(re));
    }


    /// New4 constructs a new QRegularExpression object.
    pub fn New4(pattern: []const u8, options: i32) QtC.QRegularExpression {
        const pattern_str = qtc.libqt_string{
    .len = pattern.len,
    .data = pattern.ptr,
};
return qtc.QRegularExpression_new4(pattern_str, @intCast(options));
    }


    pub fn PatternOptions(self: ?*anyopaque, ) i32 {
        return qtc.QRegularExpression_PatternOptions(@ptrCast(self));
    }

    pub fn SetPatternOptions(self: ?*anyopaque, options: i32) void {
        qtc.QRegularExpression_SetPatternOptions(@ptrCast(self), @intCast(options));
    }

    pub fn OperatorAssign(self: ?*anyopaque, re: ?*anyopaque) void {
        qtc.QRegularExpression_OperatorAssign(@ptrCast(self), @ptrCast(re));
    }

    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QRegularExpression_Swap(@ptrCast(self), @ptrCast(other));
    }

    pub fn Pattern(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QRegularExpression_Pattern(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qregularexpression.Pattern: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetPattern(self: ?*anyopaque, pattern: []const u8) void {
        const pattern_str = qtc.libqt_string{
    .len = pattern.len,
    .data = pattern.ptr,
};
qtc.QRegularExpression_SetPattern(@ptrCast(self), pattern_str);
    }

    pub fn IsValid(self: ?*anyopaque, ) bool {
        return qtc.QRegularExpression_IsValid(@ptrCast(self));
    }

    pub fn PatternErrorOffset(self: ?*anyopaque, ) i64 {
        return qtc.QRegularExpression_PatternErrorOffset(@ptrCast(self));
    }

    pub fn ErrorString(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QRegularExpression_ErrorString(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qregularexpression.ErrorString: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn CaptureCount(self: ?*anyopaque, ) i32 {
        return qtc.QRegularExpression_CaptureCount(@ptrCast(self));
    }

    pub fn NamedCaptureGroups(self: ?*anyopaque, ) []const u8 {
        return qtc.QRegularExpression_NamedCaptureGroups(@ptrCast(self));
    }

    pub fn Match(self: ?*anyopaque, subject: []const u8) QtC.QRegularExpressionMatch {
        const subject_str = qtc.libqt_string{
    .len = subject.len,
    .data = subject.ptr,
};
return qtc.QRegularExpression_Match(@ptrCast(self), subject_str);
    }

    pub fn GlobalMatch(self: ?*anyopaque, subject: []const u8) QtC.QRegularExpressionMatchIterator {
        const subject_str = qtc.libqt_string{
    .len = subject.len,
    .data = subject.ptr,
};
return qtc.QRegularExpression_GlobalMatch(@ptrCast(self), subject_str);
    }

    pub fn Optimize(self: ?*anyopaque, ) void {
        qtc.QRegularExpression_Optimize(@ptrCast(self));
    }

    pub fn Escape(str: []const u8, allocator: std.mem.Allocator) []const u8 {
        const str_str = qtc.libqt_string{
    .len = str.len,
    .data = str.ptr,
};
const _str = qtc.QRegularExpression_Escape(str_str);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qregularexpression.Escape: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn WildcardToRegularExpression(str: []const u8, allocator: std.mem.Allocator) []const u8 {
        const str_str = qtc.libqt_string{
    .len = str.len,
    .data = str.ptr,
};
const _str = qtc.QRegularExpression_WildcardToRegularExpression(str_str);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qregularexpression.WildcardToRegularExpression: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn AnchoredPattern(expression: []const u8, allocator: std.mem.Allocator) []const u8 {
        const expression_str = qtc.libqt_string{
    .len = expression.len,
    .data = expression.ptr,
};
const _str = qtc.QRegularExpression_AnchoredPattern(expression_str);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qregularexpression.AnchoredPattern: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Match22(self: ?*anyopaque, subject: []const u8, offset: i64) QtC.QRegularExpressionMatch {
        const subject_str = qtc.libqt_string{
    .len = subject.len,
    .data = subject.ptr,
};
return qtc.QRegularExpression_Match22(@ptrCast(self), subject_str, offset);
    }

    pub fn Match3(self: ?*anyopaque, subject: []const u8, offset: i64, matchType: i32) QtC.QRegularExpressionMatch {
        const subject_str = qtc.libqt_string{
    .len = subject.len,
    .data = subject.ptr,
};
return qtc.QRegularExpression_Match3(@ptrCast(self), subject_str, offset, @intCast(matchType));
    }

    pub fn Match4(self: ?*anyopaque, subject: []const u8, offset: i64, matchType: i32, matchOptions: i32) QtC.QRegularExpressionMatch {
        const subject_str = qtc.libqt_string{
    .len = subject.len,
    .data = subject.ptr,
};
return qtc.QRegularExpression_Match4(@ptrCast(self), subject_str, offset, @intCast(matchType), @intCast(matchOptions));
    }

    pub fn GlobalMatch22(self: ?*anyopaque, subject: []const u8, offset: i64) QtC.QRegularExpressionMatchIterator {
        const subject_str = qtc.libqt_string{
    .len = subject.len,
    .data = subject.ptr,
};
return qtc.QRegularExpression_GlobalMatch22(@ptrCast(self), subject_str, offset);
    }

    pub fn GlobalMatch3(self: ?*anyopaque, subject: []const u8, offset: i64, matchType: i32) QtC.QRegularExpressionMatchIterator {
        const subject_str = qtc.libqt_string{
    .len = subject.len,
    .data = subject.ptr,
};
return qtc.QRegularExpression_GlobalMatch3(@ptrCast(self), subject_str, offset, @intCast(matchType));
    }

    pub fn GlobalMatch4(self: ?*anyopaque, subject: []const u8, offset: i64, matchType: i32, matchOptions: i32) QtC.QRegularExpressionMatchIterator {
        const subject_str = qtc.libqt_string{
    .len = subject.len,
    .data = subject.ptr,
};
return qtc.QRegularExpression_GlobalMatch4(@ptrCast(self), subject_str, offset, @intCast(matchType), @intCast(matchOptions));
    }

    pub fn WildcardToRegularExpression22(str: []const u8, options: i32, allocator: std.mem.Allocator) []const u8 {
        const str_str = qtc.libqt_string{
    .len = str.len,
    .data = str.ptr,
};
const _str = qtc.QRegularExpression_WildcardToRegularExpression22(str_str, @intCast(options));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qregularexpression.WildcardToRegularExpression22: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QRegularExpression_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qregularexpressionmatch.html
pub const qregularexpressionmatch = struct {

    /// New constructs a new QRegularExpressionMatch object.
    pub fn New() QtC.QRegularExpressionMatch {
        return qtc.QRegularExpressionMatch_new();
    }


    /// New2 constructs a new QRegularExpressionMatch object.
    pub fn New2(match: ?*anyopaque) QtC.QRegularExpressionMatch {
        return qtc.QRegularExpressionMatch_new2(@ptrCast(match));
    }


    pub fn OperatorAssign(self: ?*anyopaque, match: ?*anyopaque) void {
        qtc.QRegularExpressionMatch_OperatorAssign(@ptrCast(self), @ptrCast(match));
    }

    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QRegularExpressionMatch_Swap(@ptrCast(self), @ptrCast(other));
    }

    pub fn RegularExpression(self: ?*anyopaque, ) QtC.QRegularExpression {
        return qtc.QRegularExpressionMatch_RegularExpression(@ptrCast(self));
    }

    pub fn MatchType(self: ?*anyopaque, ) i32 {
        return qtc.QRegularExpressionMatch_MatchType(@ptrCast(self));
    }

    pub fn MatchOptions(self: ?*anyopaque, ) i32 {
        return qtc.QRegularExpressionMatch_MatchOptions(@ptrCast(self));
    }

    pub fn HasMatch(self: ?*anyopaque, ) bool {
        return qtc.QRegularExpressionMatch_HasMatch(@ptrCast(self));
    }

    pub fn HasPartialMatch(self: ?*anyopaque, ) bool {
        return qtc.QRegularExpressionMatch_HasPartialMatch(@ptrCast(self));
    }

    pub fn IsValid(self: ?*anyopaque, ) bool {
        return qtc.QRegularExpressionMatch_IsValid(@ptrCast(self));
    }

    pub fn LastCapturedIndex(self: ?*anyopaque, ) i32 {
        return qtc.QRegularExpressionMatch_LastCapturedIndex(@ptrCast(self));
    }

    pub fn HasCaptured2(self: ?*anyopaque, nth: i32) bool {
        return qtc.QRegularExpressionMatch_HasCaptured2(@ptrCast(self), nth);
    }

    pub fn Captured(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QRegularExpressionMatch_Captured(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qregularexpressionmatch.Captured: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn CapturedTexts(self: ?*anyopaque, ) []const u8 {
        return qtc.QRegularExpressionMatch_CapturedTexts(@ptrCast(self));
    }

    pub fn CapturedStart(self: ?*anyopaque, ) i64 {
        return qtc.QRegularExpressionMatch_CapturedStart(@ptrCast(self));
    }

    pub fn CapturedLength(self: ?*anyopaque, ) i64 {
        return qtc.QRegularExpressionMatch_CapturedLength(@ptrCast(self));
    }

    pub fn CapturedEnd(self: ?*anyopaque, ) i64 {
        return qtc.QRegularExpressionMatch_CapturedEnd(@ptrCast(self));
    }

    pub fn Captured1(self: ?*anyopaque, nth: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QRegularExpressionMatch_Captured1(@ptrCast(self), nth);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qregularexpressionmatch.Captured1: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn CapturedStart1(self: ?*anyopaque, nth: i32) i64 {
        return qtc.QRegularExpressionMatch_CapturedStart1(@ptrCast(self), nth);
    }

    pub fn CapturedLength1(self: ?*anyopaque, nth: i32) i64 {
        return qtc.QRegularExpressionMatch_CapturedLength1(@ptrCast(self), nth);
    }

    pub fn CapturedEnd1(self: ?*anyopaque, nth: i32) i64 {
        return qtc.QRegularExpressionMatch_CapturedEnd1(@ptrCast(self), nth);
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QRegularExpressionMatch_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qregularexpressionmatchiterator.html
pub const qregularexpressionmatchiterator = struct {

    /// New constructs a new QRegularExpressionMatchIterator object.
    pub fn New() QtC.QRegularExpressionMatchIterator {
        return qtc.QRegularExpressionMatchIterator_new();
    }


    /// New2 constructs a new QRegularExpressionMatchIterator object.
    pub fn New2(iterator: ?*anyopaque) QtC.QRegularExpressionMatchIterator {
        return qtc.QRegularExpressionMatchIterator_new2(@ptrCast(iterator));
    }


    pub fn OperatorAssign(self: ?*anyopaque, iterator: ?*anyopaque) void {
        qtc.QRegularExpressionMatchIterator_OperatorAssign(@ptrCast(self), @ptrCast(iterator));
    }

    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QRegularExpressionMatchIterator_Swap(@ptrCast(self), @ptrCast(other));
    }

    pub fn IsValid(self: ?*anyopaque, ) bool {
        return qtc.QRegularExpressionMatchIterator_IsValid(@ptrCast(self));
    }

    pub fn HasNext(self: ?*anyopaque, ) bool {
        return qtc.QRegularExpressionMatchIterator_HasNext(@ptrCast(self));
    }

    pub fn Next(self: ?*anyopaque, ) QtC.QRegularExpressionMatch {
        return qtc.QRegularExpressionMatchIterator_Next(@ptrCast(self));
    }

    pub fn PeekNext(self: ?*anyopaque, ) QtC.QRegularExpressionMatch {
        return qtc.QRegularExpressionMatchIterator_PeekNext(@ptrCast(self));
    }

    pub fn RegularExpression(self: ?*anyopaque, ) QtC.QRegularExpression {
        return qtc.QRegularExpressionMatchIterator_RegularExpression(@ptrCast(self));
    }

    pub fn MatchType(self: ?*anyopaque, ) i32 {
        return qtc.QRegularExpressionMatchIterator_MatchType(@ptrCast(self));
    }

    pub fn MatchOptions(self: ?*anyopaque, ) i32 {
        return qtc.QRegularExpressionMatchIterator_MatchOptions(@ptrCast(self));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QRegularExpressionMatchIterator_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/regularexpression.html#types
pub const enums = struct {
    pub const PatternOption = enum {
        pub const NoPatternOption: i32 = 0;
        pub const CaseInsensitiveOption: i32 = 1;
        pub const DotMatchesEverythingOption: i32 = 2;
        pub const MultilineOption: i32 = 4;
        pub const ExtendedPatternSyntaxOption: i32 = 8;
        pub const InvertedGreedinessOption: i32 = 16;
        pub const DontCaptureOption: i32 = 32;
        pub const UseUnicodePropertiesOption: i32 = 64;
    };

    pub const MatchType = enum {
        pub const NormalMatch: i32 = 0;
        pub const PartialPreferCompleteMatch: i32 = 1;
        pub const PartialPreferFirstMatch: i32 = 2;
        pub const NoMatch: i32 = 3;
    };

    pub const MatchOption = enum {
        pub const NoMatchOption: i32 = 0;
        pub const AnchorAtOffsetMatchOption: i32 = 1;
        pub const AnchoredMatchOption: i32 = 1;
        pub const DontCheckSubjectStringMatchOption: i32 = 2;
    };

    pub const WildcardConversionOption = enum {
        pub const DefaultWildcardConversion: i32 = 0;
        pub const UnanchoredWildcardConversion: i32 = 1;
        pub const NonPathWildcardConversion: i32 = 2;
    };

};

File: binding/src/libqregularexpression.cpp
#include <QRegularExpression>
#include <QRegularExpressionMatch>
#include <QRegularExpressionMatchIterator>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <qregularexpression.h>
#include "libqregularexpression.h"
#include "libqregularexpression.hxx"

QRegularExpression* QRegularExpression_new() {
	 return new QRegularExpression();
}

QRegularExpression* QRegularExpression_new2(const libqt_string pattern) {
	 return new QRegularExpression(QString::fromUtf8(pattern.data, pattern.len));
}

QRegularExpression* QRegularExpression_new3(const QRegularExpression* re) {
	 return new QRegularExpression(*re);
}

QRegularExpression* QRegularExpression_new4(const libqt_string pattern, int options) {
	 return new QRegularExpression(QString::fromUtf8(pattern.data, pattern.len), static_cast<QFlags<QRegularExpression::PatternOption>>(options));
}

int QRegularExpression_PatternOptions(const QRegularExpression* self) {
	return self->patternOptions();
}

void QRegularExpression_SetPatternOptions(QRegularExpression* self, int options) {
	self->setPatternOptions(static_cast<QFlags<QRegularExpression::PatternOption>>(options));
}

void QRegularExpression_OperatorAssign(QRegularExpression* self, const QRegularExpression* re) {
	self->operator=(*re);
}

void QRegularExpression_Swap(QRegularExpression* self, QRegularExpression* other) {
	self->swap(*other);
}

libqt_string QRegularExpression_Pattern(const QRegularExpression* self) {
	QString _ret = self->pattern();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QRegularExpression_SetPattern(QRegularExpression* self, const libqt_string pattern) {
	self->setPattern(QString::fromUtf8(pattern.data, pattern.len));
}

bool QRegularExpression_IsValid(const QRegularExpression* self) {
	return self->isValid();
}

ptrdiff_t QRegularExpression_PatternErrorOffset(const QRegularExpression* self) {
	return self->patternErrorOffset();
}

libqt_string QRegularExpression_ErrorString(const QRegularExpression* self) {
	QString _ret = self->errorString();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

int QRegularExpression_CaptureCount(const QRegularExpression* self) {
	return self->captureCount();
}

libqt_list QRegularExpression_NamedCaptureGroups(const QRegularExpression* self) {
	auto _ret = self->namedCaptureGroups();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		QByteArray _b = _ret[_i].toUtf8();
		libqt_string* _str = new libqt_string();
		_str->len = _b.length();
		_str->data = static_cast<const char*>(malloc(_str->len + 1));
		memcpy((void*)_str->data, _b.data(), _str->len);
		((char*)_str->data)[_str->len] = '\0';
		_data[_i] = _str;
	}
	return _arr;
}

QRegularExpressionMatch* QRegularExpression_Match(const QRegularExpression* self, const libqt_string subject) {
	return new QRegularExpressionMatch(self->match(QString::fromUtf8(subject.data, subject.len)));
}

QRegularExpressionMatchIterator* QRegularExpression_GlobalMatch(const QRegularExpression* self, const libqt_string subject) {
	return new QRegularExpressionMatchIterator(self->globalMatch(QString::fromUtf8(subject.data, subject.len)));
}

void QRegularExpression_Optimize(const QRegularExpression* self) {
	self->optimize();
}

libqt_string QRegularExpression_Escape(const libqt_string str) {
	QString _ret = QRegularExpression::escape(QString::fromUtf8(str.data, str.len));
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QRegularExpression_WildcardToRegularExpression(const libqt_string str) {
	QString _ret = QRegularExpression::wildcardToRegularExpression(QString::fromUtf8(str.data, str.len));
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QRegularExpression_AnchoredPattern(const libqt_string expression) {
	QString _ret = QRegularExpression::anchoredPattern(QString::fromUtf8(expression.data, expression.len));
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QRegularExpressionMatch* QRegularExpression_Match22(const QRegularExpression* self, const libqt_string subject, ptrdiff_t offset) {
	return new QRegularExpressionMatch(self->match(QString::fromUtf8(subject.data, subject.len), offset));
}

QRegularExpressionMatch* QRegularExpression_Match3(const QRegularExpression* self, const libqt_string subject, ptrdiff_t offset, int matchType) {
	return new QRegularExpressionMatch(self->match(QString::fromUtf8(subject.data, subject.len), offset, static_cast<QRegularExpression::MatchType>(matchType)));
}

QRegularExpressionMatch* QRegularExpression_Match4(const QRegularExpression* self, const libqt_string subject, ptrdiff_t offset, int matchType, int matchOptions) {
	return new QRegularExpressionMatch(self->match(QString::fromUtf8(subject.data, subject.len), offset, static_cast<QRegularExpression::MatchType>(matchType), static_cast<QFlags<QRegularExpression::MatchOption>>(matchOptions)));
}

QRegularExpressionMatchIterator* QRegularExpression_GlobalMatch22(const QRegularExpression* self, const libqt_string subject, ptrdiff_t offset) {
	return new QRegularExpressionMatchIterator(self->globalMatch(QString::fromUtf8(subject.data, subject.len), offset));
}

QRegularExpressionMatchIterator* QRegularExpression_GlobalMatch3(const QRegularExpression* self, const libqt_string subject, ptrdiff_t offset, int matchType) {
	return new QRegularExpressionMatchIterator(self->globalMatch(QString::fromUtf8(subject.data, subject.len), offset, static_cast<QRegularExpression::MatchType>(matchType)));
}

QRegularExpressionMatchIterator* QRegularExpression_GlobalMatch4(const QRegularExpression* self, const libqt_string subject, ptrdiff_t offset, int matchType, int matchOptions) {
	return new QRegularExpressionMatchIterator(self->globalMatch(QString::fromUtf8(subject.data, subject.len), offset, static_cast<QRegularExpression::MatchType>(matchType), static_cast<QFlags<QRegularExpression::MatchOption>>(matchOptions)));
}

libqt_string QRegularExpression_WildcardToRegularExpression22(const libqt_string str, int options) {
	QString _ret = QRegularExpression::wildcardToRegularExpression(QString::fromUtf8(str.data, str.len), static_cast<QFlags<QRegularExpression::WildcardConversionOption>>(options));
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QRegularExpression_Delete(QRegularExpression* self) {
    delete self;
}

QRegularExpressionMatch* QRegularExpressionMatch_new() {
	 return new QRegularExpressionMatch();
}

QRegularExpressionMatch* QRegularExpressionMatch_new2(const QRegularExpressionMatch* match) {
	 return new QRegularExpressionMatch(*match);
}

void QRegularExpressionMatch_OperatorAssign(QRegularExpressionMatch* self, const QRegularExpressionMatch* match) {
	self->operator=(*match);
}

void QRegularExpressionMatch_Swap(QRegularExpressionMatch* self, QRegularExpressionMatch* other) {
	self->swap(*other);
}

QRegularExpression* QRegularExpressionMatch_RegularExpression(const QRegularExpressionMatch* self) {
	return new QRegularExpression(self->regularExpression());
}

int QRegularExpressionMatch_MatchType(const QRegularExpressionMatch* self) {
	return self->matchType();
}

int QRegularExpressionMatch_MatchOptions(const QRegularExpressionMatch* self) {
	return self->matchOptions();
}

bool QRegularExpressionMatch_HasMatch(const QRegularExpressionMatch* self) {
	return self->hasMatch();
}

bool QRegularExpressionMatch_HasPartialMatch(const QRegularExpressionMatch* self) {
	return self->hasPartialMatch();
}

bool QRegularExpressionMatch_IsValid(const QRegularExpressionMatch* self) {
	return self->isValid();
}

int QRegularExpressionMatch_LastCapturedIndex(const QRegularExpressionMatch* self) {
	return self->lastCapturedIndex();
}

bool QRegularExpressionMatch_HasCaptured2(const QRegularExpressionMatch* self, int nth) {
	return self->hasCaptured(nth);
}

libqt_string QRegularExpressionMatch_Captured(const QRegularExpressionMatch* self) {
	QString _ret = self->captured();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_list QRegularExpressionMatch_CapturedTexts(const QRegularExpressionMatch* self) {
	auto _ret = self->capturedTexts();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		QByteArray _b = _ret[_i].toUtf8();
		libqt_string* _str = new libqt_string();
		_str->len = _b.length();
		_str->data = static_cast<const char*>(malloc(_str->len + 1));
		memcpy((void*)_str->data, _b.data(), _str->len);
		((char*)_str->data)[_str->len] = '\0';
		_data[_i] = _str;
	}
	return _arr;
}

ptrdiff_t QRegularExpressionMatch_CapturedStart(const QRegularExpressionMatch* self) {
	return self->capturedStart();
}

ptrdiff_t QRegularExpressionMatch_CapturedLength(const QRegularExpressionMatch* self) {
	return self->capturedLength();
}

ptrdiff_t QRegularExpressionMatch_CapturedEnd(const QRegularExpressionMatch* self) {
	return self->capturedEnd();
}

libqt_string QRegularExpressionMatch_Captured1(const QRegularExpressionMatch* self, int nth) {
	QString _ret = self->captured(nth);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

ptrdiff_t QRegularExpressionMatch_CapturedStart1(const QRegularExpressionMatch* self, int nth) {
	return self->capturedStart(nth);
}

ptrdiff_t QRegularExpressionMatch_CapturedLength1(const QRegularExpressionMatch* self, int nth) {
	return self->capturedLength(nth);
}

ptrdiff_t QRegularExpressionMatch_CapturedEnd1(const QRegularExpressionMatch* self, int nth) {
	return self->capturedEnd(nth);
}

void QRegularExpressionMatch_Delete(QRegularExpressionMatch* self) {
    delete self;
}

QRegularExpressionMatchIterator* QRegularExpressionMatchIterator_new() {
	 return new QRegularExpressionMatchIterator();
}

QRegularExpressionMatchIterator* QRegularExpressionMatchIterator_new2(const QRegularExpressionMatchIterator* iterator) {
	 return new QRegularExpressionMatchIterator(*iterator);
}

void QRegularExpressionMatchIterator_OperatorAssign(QRegularExpressionMatchIterator* self, const QRegularExpressionMatchIterator* iterator) {
	self->operator=(*iterator);
}

void QRegularExpressionMatchIterator_Swap(QRegularExpressionMatchIterator* self, QRegularExpressionMatchIterator* other) {
	self->swap(*other);
}

bool QRegularExpressionMatchIterator_IsValid(const QRegularExpressionMatchIterator* self) {
	return self->isValid();
}

bool QRegularExpressionMatchIterator_HasNext(const QRegularExpressionMatchIterator* self) {
	return self->hasNext();
}

QRegularExpressionMatch* QRegularExpressionMatchIterator_Next(QRegularExpressionMatchIterator* self) {
	return new QRegularExpressionMatch(self->next());
}

QRegularExpressionMatch* QRegularExpressionMatchIterator_PeekNext(const QRegularExpressionMatchIterator* self) {
	return new QRegularExpressionMatch(self->peekNext());
}

QRegularExpression* QRegularExpressionMatchIterator_RegularExpression(const QRegularExpressionMatchIterator* self) {
	return new QRegularExpression(self->regularExpression());
}

int QRegularExpressionMatchIterator_MatchType(const QRegularExpressionMatchIterator* self) {
	return self->matchType();
}

int QRegularExpressionMatchIterator_MatchOptions(const QRegularExpressionMatchIterator* self) {
	return self->matchOptions();
}

void QRegularExpressionMatchIterator_Delete(QRegularExpressionMatchIterator* self) {
    delete self;
}

